OpenRA.Game/Server/OrderBuffer.cs

```diff
@@ -0,0 +1,132 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2021 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+
+namespace OpenRA.Server
+{
+	public class OrderBuffer
+	{
+		const int NumberOfFrames = 20;
+		const int Interval = 1000;
+
+		// Limit the TickScale to maximum of 10%
+		const float MaxTickScale = 1.1f;
+
+		Stopwatch gameTimer;
+		long nextUpdate = 0;
+
+		readonly ConcurrentDictionary<int, long> timestamps = new ConcurrentDictionary<int, long>();
+		readonly ConcurrentDictionary<int, Queue<long>> deltas = new ConcurrentDictionary<int, Queue<long>>();
+
+		int timestep;
+		int ticksPerSecond;
+		int baselinePlayer;
+		List<int> players;
+
+		public void AddOrderTimestamp(int playerIndex)
+		{
+			timestamps[playerIndex] = gameTimer.ElapsedMilliseconds;
+
+			if (timestamps.Values.All(t => t > 0))
```

Comment:
In a rare condition one player - i.e. hosting the server might add a order timestamp at - within - millisecond `0`. From there on timestamp delay between that one player and the others will always be maximum. Causing the game to run slow contineously. 

` -1` might be a saver nil value.



`-1` may be a saver null value.